### PR TITLE
Add short-circuit evaluation boolean syntax  

### LINKED ISSUE #3362: Boolean operators and/or are evaluated eagerly (add lazy &&/||)
### What did you expect to see?
Boolean operators should lazy evaluate operands.

**UPDATE: Instead if changing behavior of existing `and` and `or` boolean operators, lazy variants will be added `&&` and `||`.**

### What did you see instead?
The error _Error: Method `get` is not defined on String._ is reported instead.

### Steps to reproduce the bug
`if(false and ("x".get(0))) {}` should not throw error because the first operand `false` makes the whole expression false.

https://github.com/rchain/rchain/blob/09d9acc916e008dc93a7289202262fd5a4d9d41d/rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala#L605-L617

Related issue: #3344

### PR DESCRIPTION
**Pull request by zsluedem**
## Overview

https://github.com/rchain/rchain/issues/3362

Add `&&` and `||` syntax for rholang which support short circuit boolean evaluation. 

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ShortCircuitBooleanSpec.scala [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        s"""
+           # @"${outcomeCh}"!(1<0 || [1,2].nth(-1))
+           # """.stripMargin('#')
+
+      execute(term2).runSyncUnsafe(maxDuration) should equal(Left(ReduceError(reduceErrorMsg)))
+    }
+  }
+
+  "&&" should {
+    "work good with `and`" in {
```
Comment (tgrospic) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/ShortCircuitBooleanSpec.scala]: _Work good_ is very general description so there is no indication what is tested.

**Review by tgrospic — APPROVED** [review]
Nice work!

Please create the same PR for `rchain:feature/v0.12-rpc` branch and link PR in the epic #3642.

**Review by hilltracer — APPROVED** [review]
https://github.com/rchain/rchain/pull/3652#pullrequestreview-930936407

**Review by stanislavlyalin — APPROVED** [review]
LGTM. Good division into commits!

**Comment by zsluedem** [comment]
bors merge

**Comment by zsluedem** [comment]
bors merge

**Pull request merged**